Enable quantized_max_pool3d

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101654

**Summary**
Enable `quantized_max_pool3d` kernel to fix the issue https://github.com/pytorch/pytorch/issues/101386.

**Test Plan**
```
clear && python -u -m pytest -s -v test_quantized_op.py -k test_max_pool3d
clear && python -u -m pytest -s -v test_quantized_op.py -k test_max_pool3d_nhwc
```


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

leslie-fang-intel committed:
Enable quantized_max_pool3d

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable quantized_max_pool3d"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable quantized_max_pool3d"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable quantized_max_pool3d"


**Summary**
Enable `quantized_max_pool3d` kernel to fix the issue https://github.com/pytorch/pytorch/issues/101386.

**Test Plan**
```
clear && python -u -m pytest -s -v test_quantized_op.py -k test_max_pool3d
clear && python -u -m pytest -s -v test_quantized_op.py -k test_max_pool3d_nhwc
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable quantized_max_pool3d"


**Summary**
Enable `quantized_max_pool3d` kernel to fix the issue https://github.com/pytorch/pytorch/issues/101386.

**Test Plan**
```
clear && python -u -m pytest -s -v test_quantized_op.py -k test_max_pool3d
clear && python -u -m pytest -s -v test_quantized_op.py -k test_max_pool3d_nhwc
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable quantized_max_pool3d"


**Summary**
Enable `quantized_max_pool3d` kernel to fix the issue https://github.com/pytorch/pytorch/issues/101386.

**Test Plan**
```
clear && python -u -m pytest -s -v test_quantized_op.py -k test_max_pool3d
clear && python -u -m pytest -s -v test_quantized_op.py -k test_max_pool3d_nhwc
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

albanD reviewed (changes requested):
We shouldn't add a new native function for this and I think we can just use the regular dispatcher!

jgong5 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3655,6 +3655,11 @@
     QuantizedCUDA: quantized_max_pool2d_cudnn
   autogen: quantized_max_pool2d.out
 
+- func: quantized_max_pool3d(Tensor self, int[3] kernel_size, int[3] stride=[], int[3] padding=0, int[3] dilation=1, bool ceil_mode=False) -> Tensor
```
On the other hand, there are already `quantized_max_pool2d` and `quantized_max_pool1d`. Adding `quantized_max_pool3d` in parallel seems also consistent with them.

albanD replied:
hmm that shouldn't @bdhirsh are the Quantized key handled differently here?

bdhirsh replied:
hmm @albanD I think I know what's going on. With the per-backend-functionality dispatch key changes, a tensor with the `QuantizedCPU` dispatch key automatically gets the `AutogradCPU` dispatch key too ([code](https://github.com/pytorch/pytorch/blob/ac1cf00085b30eadd164ccf02e5208a59ec1b38b/c10/core/DispatchKeySet.h#L809))

So what we care about is "what gets registered to the AutogradCPU key" in the dispatcher. But... we want the composite kernel to be registered to the AutogradCPU key! Since the normal CPU backend should run that decomp.

I think there are 3 options:

(1) Add this new quantized-specific op

(2) Try to get the quantized backends to use the AutogradOther dispatch key again, instead of using AutogradCPU. I think that this would fix the problem? But I'd need to think more about what the blast radius of that change would be. wdyt?

(3) Ask users to run this quantized logic under `torch.inference_mode()`. I think this should work (in that it would successfully prevent the autograd key from running, so we hit the quantized kernel instead of the decomp), but this might not be a reasonable ask, since it means you'll get wrong results if you forget to use inference mode.

leslie-fang-intel replied:
Thanks for the comments. I think option1 as _Add this new quantized-specific op_ is what we used to do in similar kernels of `quantized_max_pool2d` and `quantized_max_pool1d`. @albanD Do you think is it ok we keep this PR in this similar way until option2 is implemented?

jgong5 reviewed (approved):
LGTM except for the comments from @albanD

mingfeima commented on aten/src/ATen/native/quantized/cpu/Pooling.cpp:
```diff
@@ -232,6 +307,125 @@ Tensor q_maxpool_2d(
+    auto qxd = qx_contig.data_ptr<Q>();
+    auto qyd = qy.data_ptr<Q>();
+
+    at::parallel_for(0, nbatch, 0, [&](int64_t start, int64_t end) {
```
Oops, this is not a good practice to use nested parallism on OpenMP, e.g. first parallel on N and then parallel on C.
This would lead to low performance on some cases, for example N=2 and you have 4 CPU cores.

Try to collapse both N and C for channels first.

leslie-fang-intel replied:
Thanks for the suggestions. Collapsed N and C for the channels first case.

mingfeima commented on aten/src/ATen/native/quantized/cpu/Pooling.cpp:
```diff
@@ -232,6 +307,125 @@ Tensor q_maxpool_2d(
+
+  std::vector<int64_t> oSizes = {nbatch, oC, oT, oH, oW};
+
+  if (qx.is_contiguous(c10::MemoryFormat::ChannelsLast3d)) {
```
* do we need to consider **non batch mode**? (batch dimension is ignored, the input is THWC which is 4-d tensor)
* using `is_contiguous` would lead to some weirdo perf issues when C=1, the current CL kernel would be slower then CF when C is not long enough for vectorization. And the vector length for Qint is 32 or 64, so when C=1, C=3, CL would have worse performance than CF.

leslie-fang-intel replied:
Thanks for your remind. What's the general solutions as we can do when meet similar issues?

mingfeima commented on aten/src/ATen/native/quantized/cpu/Pooling.cpp:
```diff
@@ -232,6 +307,125 @@ Tensor q_maxpool_2d(
+      dW,
+      ")");
+  int ndim = qx.dim();
+  TORCH_CHECK(ndim == 5, "Expecting the input tensor of rank 5.");
```
non batch mode is not allowed ?

leslie-fang-intel replied:
Thanks for the suggestions. I add a TODO to enable the `non batch mode` for `quantized maxpool3d` when meet use cases.

mingfeima commented on aten/src/ATen/native/quantized/cpu/Pooling.cpp:
```diff
@@ -94,6 +95,80 @@ void spatial_dilated_max_pooling(
+    int64_t dW, // dilation
+    T* oData) { // output arrays (data and max-index)
+  // Handle each bs
+  at::parallel_for(0, iC, 0, [&](int64_t start, int64_t end) {
```
One more thing, if we care about channels first performance. Even parallel on N*C is not enough, say N=1, C=1 and #cores = 4.

if we do not care about CF perf, that's fine.
if we do care, then parallel on NCTH and cache the output indices from W, going to be much faster.

leslie-fang-intel replied:
Thanks for the suggestions. Add a TODO in `spatial_dilated_max_pooling3d` as what we can do to further the CF performance.

mingfeima reviewed (changes requested):
The nested parallel issue for Channels First kernel is a must fix !
For other issues, you may decide whether to fix them depending on real requests from our workloads.

But be cautious that when C is not long enough for vectorization, CL might be slower for CF.
We can into similar issues with `interpolation kernels` or upsampling, the issue should be even worse on Qint since the vector length is larger.

leslie-fang-intel committed:
Update on "Enable quantized_max_pool3d"


**Summary**
Enable `quantized_max_pool3d` kernel to fix the issue https://github.com/pytorch/pytorch/issues/101386.

**Test Plan**
```
clear && python -u -m pytest -s -v test_quantized_op.py -k test_max_pool3d
clear && python -u -m pytest -s -v test_quantized_op.py -k test_max_pool3d_nhwc
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
> The nested parallel issue for Channels First kernel is a must fix ! For other issues, you may decide whether to fix them depending on real requests from our workloads.
> 
> But be cautious that when C is not long enough for vectorization, CL might be slower for CF. We can into similar issues with `interpolation kernels` or upsampling, the issue should be even worse on Qint since the vector length is larger.

Hi @mingfeima, Thanks for the suggestions.

- The nested parallel issue has been fixed.
- I have added TODO with your suggestions in kernel implementation for further performance optimization when meet real workloads requests. 
- As for the issue that CL might be slower for CF when C is not long enough, what's the general solutions we can do when meet similar issues?

albanD reviewed (approved):
This sounds ok for this one. But we must fix the registration before the next op is being added.
With the new compiler backends using the aten API as the de-facto public API, we need to be a lot stricter than we used to be on what we do there!

leslie-fang-intel committed:
Update on "Enable quantized_max_pool3d"


**Summary**
Enable `quantized_max_pool3d` kernel to fix the issue https://github.com/pytorch/pytorch/issues/101386.

**Test Plan**
```
clear && python -u -m pytest -s -v test_quantized_op.py -k test_max_pool3d
clear && python -u -m pytest -s -v test_quantized_op.py -k test_max_pool3d_nhwc
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
